linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CustomJSHelperBuilder.scala

```diff
@@ -0,0 +1,284 @@
+ *  ...
+ *  }}}
+ *
+ *  In addition to the input management, `CustomJSHelperBuilder` provides tools
+ *  for local name management, and generation of `js.ParamDef`s for inner
+ *  functions.
+ */
+class CustomJSHelperBuilder()(implicit ctx: WasmContext, pos: Position) {
+  import CustomJSHelperBuilder._
+
+  private val usedGlobalRefs = mutable.Set.empty[String]
+  private val allocatedLocalIdentResolvers = mutable.ListBuffer.empty[LocalResolver]
+
+  private val jsParamDefs = mutable.ListBuffer.empty[js.ParamDef]
+  private val wasmParamTypes = mutable.ListBuffer.empty[watpe.Type]
+
+  def newLocalIdent(origName: String): js.DelayedIdent = {
+    val resolver = new LocalResolver(origName)
+    allocatedLocalIdentResolvers += resolver
+    js.DelayedIdent(resolver)
+  }
+
+  def newLocalIdent(name: LocalName): js.DelayedIdent =
+    newLocalIdent(ctx.jsNameGen.genName(name))
+
+  private def addParamDef(origName: String, wasmType: watpe.Type): js.VarRef = {
+    val jsParamDef = js.ParamDef(newLocalIdent(origName))
+    jsParamDefs += jsParamDef
+    wasmParamTypes += wasmType
+    jsParamDef.ref
+  }
+
+  /** Adds an input of an arbitrary Wasm type, to be evaluated on the Wasm
+   *  stack.
+   *
+   *  The `evalValue` must add code to the call site context to evaluate the
+   *  value on the Wasm site. It is passed as by-name parameter to show
+   *  intent, but is in fact called immediately.
+   *
+   *  @return
+   *    A `js.VarRef` that can be used in the JS helper to read the input.
+   */
+  def addWasmInput(origName: String, wasmType: watpe.Type)(evalValue: => Unit): js.VarRef = {
+    evalValue
+    addParamDef(origName, wasmType)
+  }
+
+  def genGlobalRef(name: String): js.VarRef = {
+    usedGlobalRefs += name
+    js.VarRef(js.Ident(name))
+  }
+
+  def genJSNativeLoadSpec(jsNativeLoadSpec: JSNativeLoadSpec): js.Tree = {
+    def genFollowPath(owner: js.Tree, path: List[String]): js.Tree = {
+      path.foldLeft(owner) { (owner, item) =>
+        js.BracketSelect.makeOptimized(owner, js.StringLiteral(item))
+      }
+    }
+
+    jsNativeLoadSpec match {
+      case JSNativeLoadSpec.Global(globalRef, path) =>
+        genFollowPath(genGlobalRef(globalRef), path)
+      case JSNativeLoadSpec.Import(module, path) =>
+        genFollowPath(js.VarRef(js.Ident("imported" + JSNameGen.genModuleName(module))), path)
+      case JSNativeLoadSpec.ImportWithGlobalFallback(importSpec, _) =>
+        genJSNativeLoadSpec(importSpec)
+    }
+  }
+
+  def genJSParamDef(param: ParamDef): js.ParamDef =
+    js.ParamDef(newLocalIdent(param.name.name))
+
+  def genJSParamDefs(params: List[ParamDef],
+      restParam: Option[ParamDef]): (List[js.ParamDef], Option[js.ParamDef]) = {
+    (params.map(genJSParamDef(_)), restParam.map(genJSParamDef(_)))
+  }
+
+  def build(resultType: Type)(body: js.Tree): wanme.FunctionID = {
+    // Allocate names for the local ident resolvers
+    val usedLocalNames = usedGlobalRefs.clone()
```

Comment:
Rather than cloning, use an immutable set with var and structural sharing? Or are you concerned about performance?

Reply:
Not really concerned either way, but our fresh name generators have historically all used mutable sets. So IMO the `clone()` is more consistent with the rest of our codebase.